modules/dnn/src/layers/convolution_layer.cpp

```diff
@@ -256,21 +268,23 @@ class ConvolutionLayerImpl CV_FINAL : public BaseConvolutionLayerImpl
 
         internals.clear();
 
-        int inpCn = inputs[0][1];
-        int inpH = inputs[0][2];
-        int inpW = inputs[0][3];
+        CV_Assert(inputs.size() != 0);
+        std::vector<int> inpShape;
+        for (int i = 2; i < inputs[0].size(); i++) {
+            inpShape.push_back(inputs[0][i]);
+        }
 
+        std::vector<int> outShape;
         int outCn = blobs[0].size[0];
-        Size out;
-
+        int inpCn = inputs[0][1];
         if (padMode.empty())
         {
-            out.height = (inpH + 2 * pad.height - (dilation.height * (kernel.height - 1) + 1)) / stride.height + 1;
-            out.width = (inpW + 2 * pad.width - (dilation.width * (kernel.width - 1) + 1)) / stride.width + 1;
+            for (int i = 0; i < inpShape.size(); i++)
+                outShape.push_back((inpShape[i] + pads_begin[i] + pads_end[i] - dilations[i] * (kernel_size[i] - 1) - 1) / strides[i] + 1);
         }
         else
         {
-            getConvPoolOutParams(Size(inpW, inpH), kernel, stride, padMode, dilation, out);
+            getConvPoolOutParams(inpShape, kernel_size, strides, padMode, dilations, outShape);
```

Comment:
I think that we can miss the case with `padMode.empty()` for 3D convolutions here because all the arguments are casted to `cv::Size` which has only two values.

I think it's better to modify `getConvPoolOutParams` so it receives all the spatial dimensions and can manage `padMode.empty()`:

https://github.com/opencv/opencv/blob/e6171d17f8b22163997487b16762d09671a68597/modules/dnn/src/layers/layers_common.cpp#L195-L208

Reply:
If combine these cases, there will be problems in pooling layer
`            if ((pads_end.size() == 2 && (pads_end[0] || pads_end[1])) ||
                (pads_end.size() == 3 && (pads_end[0] || pads_end[1] || pads_end[2])))
            {
                // If we have padding, ensure that the last pooling starts strictly
                // inside the image (instead of at the padding); otherwise clip the last.
                for (int i = 0; i < outShape.size(); i++) {
                    if ((outShape[i] - 1) * strides[i] >= inpShape[i] + pads_end[i]) {
                        --outShape[i];
                        CV_Assert((outShape[i] - 1) * strides[i] < inpShape[i] + pads_end[i]);
                    }
                }
            }`